File: docs/research/corpus-examples/search-export/python/vulnerable-3-altered.py
# MANUFACTURED per docs/VULN_CORPUS_PAIR_MANUFACTURING_PLAN.md's "Per-pair
# mechanics" (Phase 3, final batch, group 4 of 10). Genuinely distinct third
# variant from vulnerable-1.py/idiomatic-1.py's Django-ORM-escape-hatch shape
# (QuerySet.extra(where=...) vs .filter(Q(...))): a raw psycopg2 driver-level
# order/export endpoint, with no ORM involved at all -- f-string
# interpolation directly into the SQL text passed to cursor.execute().
import psycopg2


def export_orders(conn, search: str, sort_by: str) -> list:
    cursor = conn.cursor()
    # VULNERABLE: both `search` and `sort_by` are interpolated directly into
    # the SQL text via an f-string before it ever reaches cursor.execute() --
    # psycopg2's own documentation is explicit that this bypasses its driver-
    # level parameter substitution entirely, and is not equivalent to using
    # `%s` placeholders even though the connection/cursor objects are
    # identical in both files.
    query = f"""
        SELECT id, customer_name, total, created_at FROM orders
        WHERE customer_name LIKE '%{search}%'
        ORDER BY {sort_by}
    """
    cursor.execute(query)
    return cursor.fetchall()

File: docs/research/corpus-examples/search-export/node/vulnerable-3-altered.js
// MANUFACTURED per docs/VULN_CORPUS_PAIR_MANUFACTURING_PLAN.md's "Per-pair
// mechanics" (Phase 3, final batch, group 2 of 10). Closes this cell's
// remaining CWE-89 gap (distinct from vulnerable-2.js/idiomatic-2-altered.js,
// which cover this same cell's CWE-1336 SSTI-in-export group). Genuinely
// distinct third variant from vulnerable-1.js/idiomatic-1.js: a Sequelize
// ORM (not raw pg/knex) product-search endpoint, where the WHERE clause uses
// Sequelize's own parameterized operators correctly, but the ORDER BY column
// is built with sequelize.literal() from raw, unvalidated request input --
// the same "ORM query-builder misuse, not raw-concat" shape as vulnerable-1.js,
// in a different, independently common Node ORM.
const { Op } = require('sequelize');
const { sequelize, Product } = require('../models');

async function searchProducts(req, res, next) {
  try {
    const { q, category, sort_field, sort_dir } = req.query;

    const where = {};
    if (q) where.name = { [Op.like]: `%${q}%` };
    if (category) where.category = category;

    // VULNERABLE: sort_field/sort_dir come straight from the query string
    // and are interpolated into a raw SQL fragment via sequelize.literal(),
    // with no allowlist check against Product's real column names -- while
    // every other dynamic value in this same query (q, category) is bound
    // safely through Sequelize's own Op.like/equality operators.
    const products = await Product.findAll({
      where,
      order: [[sequelize.literal(`${sort_field || 'id'} ${sort_dir || 'ASC'}`)]],
    });

    res.json(products);
  } catch (err) {
    next(err);
  }
}

module.exports = { searchProducts };
